components/api_server/src/initialization/migrations.py

```diff
@@ -87,8 +93,32 @@ def sources_with_branch_parameter(report: dict):  # pragma: no feature-test-cove
                         yield source
 
 
-def replace_report(database: Database, report) -> None:  # pragma: no feature-test-cover
-    """Replace the report in the database."""
-    report_id = report["_id"]
-    del report["_id"]
-    database.reports.replace_one({"_id": report_id}, report)
+def add_source_parameter_hash(database: Database, report) -> None:  # pragma: no feature-test-cover
+    """Add source parameter hashes to the latest measurements."""
+    # Added after Quality-time v5.12.0, see https://github.com/ICTU/quality-time/issues/8736
+    for subject in report["subjects"].values():
+        for metric_uuid, metric in subject["metrics"].items():
+            latest_measurement = database.measurements.find_one(
+                filter={"metric_uuid": metric_uuid},
+                sort=[("start", pymongo.DESCENDING)],
+            )
+            if not latest_measurement:
+                continue
+            if latest_measurement.get("source_parameter_hash"):
+                continue
+            latest_measurement["source_parameter_hash"] = source_parameter_hash(metric)
+            replace_document(database.measurements, latest_measurement)
+
+
+def source_parameter_hash(metric) -> str:  # pragma: no feature-test-cover
+    """Return a hash of the source parameters that can be used to check for changed parameters."""
+    sources = [source for _, source in sorted(metric.get("sources", {}).items())]
+    source_parameters = [source.get("parameters") for source in sources]
+    return md5_hash(str(source_parameters))
```

Comment:
Can't we import the function from where it's being set, instead of re-implementing?

Reply:
We could, but it means the migration code would get coupled to the Metric model class. Since this is (was, since this morning's discussion) long-living code I decided against that. I'll add a comment.